### PR TITLE
Formplayer End of Form Navigation

### PR DESCRIPTION
**Pull request by wpride**
Implement end of form navigation for formplayer. Needed a few things:

We add the ability to store MenuSessions in much the same way we store FormSessions. Add a repository and mappings to do this.
When we start form entry, we save the MenuSession and pass its ID to the form session (new column in that table).
When form entry completes (in the submit controller) we re-instantiate the MenuSession from its ID. Then, we execute the form completion against that MenuSession and check if we have another command on the frame to perform.
If so, we get this command or entity list and return it along with the submit response.
We then save this menu session and add another route that allows the front end to perform new steps against this saved session.
Also add test coverage for this. 

cross https://github.com/dimagi/commcare-hq/pull/12718/files

### CODE REVIEW

File: src/main/java/application/AbstractBaseController.java [hunk 1]
```diff
@@ -37,6 +56,70 @@
...
+            return generateFormEntryScreen(menuSession);
+        } else {
+            MenuBean menuResponseBean;
+
+            // We're looking at a module or form menu
+            if (nextScreen instanceof MenuScreen) {
+                menuResponseBean = generateMenuScreen((MenuScreen) nextScreen, menuSession.getSessionWrapper(),
+                        menuSession.getId());
+            }
+            // We're looking at a case list or detail screen (probably)
```
Comment (benrudolph) [inline src/main/java/application/AbstractBaseController.java]: probably feels kind of scary in a comment. do you know the other cases?

File: src/main/java/application/FormController.java [hunk 2]
```diff
@@ -104,13 +107,15 @@ public FormEntryResponseBean getCurrent(@RequestBody CurrentRequestBean currentR
     @ApiOperation(value = "Submit the current form")
     @RequestMapping(value = Constants.URL_SUBMIT_FORM, method = RequestMethod.POST)
     @ResponseBody
-    public SubmitResponseBean submitForm(@RequestBody SubmitRequestBean submitRequestBean,
+    public Object submitForm(@RequestBody SubmitRequestBean submitRequestBean,
```
Comment (benrudolph) [inline src/main/java/application/FormController.java]: why `Object` now?

File: src/main/java/db/migration/V2__init_menu_session.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+                "    id text NOT NULL,\n" +
+                "    username text,\n" +
+                "    domain text,\n" +
+                "    appid text,\n" +
+                "    installreference text,\n" +
+                "    locale text,\n" +
+                "    commcaresession bytea,\n" +
+                "    CONSTRAINT menu_sessions_pkey PRIMARY KEY (id)\n" +
+                ")");
+        jdbcTemplate.execute("ALTER TABLE formplayer_sessions ADD menu_session_id text");
```
Comment (benrudolph) [inline src/main/java/db/migration/V2__init_menu_session.java]: we may want to add a constraint to this to ensure that there is a matching menu_session_id (i.e. it really is a foreign key)

File: src/main/java/db/migration/V2__init_menu_session.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * @author wspride
+ */
+public class V2__init_menu_session implements SpringJdbcMigration {
+
+    @Override
+    public void migrate(JdbcTemplate jdbcTemplate) throws Exception {
+        jdbcTemplate.execute("CREATE TABLE menu_sessions (\n" +
+                "    id text NOT NULL,\n" +
+                "    username text,\n" +
+                "    domain text,\n" +
```
Comment (benrudolph) [inline src/main/java/db/migration/V2__init_menu_session.java]: it's good practice to label these as NOT NULL if they are required. that way we never accidentally commit bad data

File: src/main/java/repo/impl/PostgresMenuSessionRepo.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+        if(sessionCount > 0){
+            String query = replaceTableName("UPDATE %s SET commcaresession = ? WHERE id = ?");
+            this.jdbcTemplate.update(query,  new Object[] {session.getCommcareSession(), session.getId()},
+                    new int[] {Types.BINARY, Types.VARCHAR});
+            return session;
+
+        }
+
+        String query = replaceTableName("INSERT into %s " +
```
Comment (benrudolph) [inline src/main/java/repo/impl/PostgresMenuSessionRepo.java]: shouldn't the insert only happen if sessionCount == 0

File: src/main/java/requests/NewFormRequest.java [hunk 6]
```diff
@@ -77,15 +77,17 @@ public NewFormRequest(NewSessionRequestBean bean, SessionRepo sessionRepo,
...
     }
 
     public NewFormSessionResponse getResponse() throws IOException {
         return new NewFormSessionResponse(formEntrySession);
     }
 
     private String getRestoreXml(){
-        return restoreService.getRestoreXml(domain, auth);
+        String xml =  restoreService.getRestoreXml(domain, auth);
+        System.out.println("Restoring XML: " + xml);
```
Comment (benrudolph) [inline src/main/java/requests/NewFormRequest.java]: print

File: src/test/java/tests/CaseTests.java [hunk 7]
```diff
@@ -45,6 +45,8 @@ public void testCases() throws Exception {
 
         caseFilterResponseBean = filterCases("requests/filter/filter_cases_5.json");
 
+        System.out.println("Got case filte response: " + caseFilterResponseBean);
```
Comment (benrudolph) [inline src/test/java/tests/CaseTests.java]: print

File: src/test/java/tests/DoubleManagementTest.java [hunk 8]
```diff
@@ -63,6 +64,21 @@ public void testDoubleForm() throws Exception {
...
+        CommandListResponseBean commandListResponseBean = mapper.readValue(mapper.writeValueAsString(submitResponseBean.getNextScreen()),
+                CommandListResponseBean.class);
+        assert commandListResponseBean.getCommands().length == 2;
+        assert commandListResponseBean.getCommands()[0].getDisplayText().equals("Update Parent");
+        JSONObject endOfFormNavResponse = sessionNavigateWithId(new String[] {"0"}, "derp");
+        NewFormSessionResponse followupFormResponse =
+                mapper.readValue(endOfFormNavResponse.toString(), NewFormSessionResponse.class);
+        assert followupFormResponse.getTree().length == 2;
+        assert followupFormResponse.getTree()[0].getAnswer().equals("David Ortiz");
+        assert followupFormResponse.getTree()[1].getAnswer().equals(40);
```
Comment (benrudolph) [inline src/test/java/tests/DoubleManagementTest.java]: 🙌

File: src/test/java/tests/DoubleManagementTest.java [hunk 9]
```diff
@@ -120,8 +136,9 @@ public void testAllPermutations() throws Exception {
         JSONObject parentResponseObject2 = sessionNavigate(new String[] {"2"}, "doublemgmt");
         EntityListResponse entityListResponse =
                 mapper.readValue(parentResponseObject2.toString(), EntityListResponse.class);
+        System.out.println("Entity List Response: " + entityListResponse);
```
Comment (benrudolph) [inline src/test/java/tests/DoubleManagementTest.java]: print

**Comment by benrudolph** [comment]
nice! 📝 🌏 

one thing i noticed is that menusession and form session seem very similar. is there any way we can formalize that with a interface or some shared utils? not quite sure, some of the logic goes over my head a bit so quite possibly not possible

File: src/main/java/application/AbstractBaseController.java (hunk 1 above)
Comment (wpride) [inline src/main/java/application/AbstractBaseController.java]: Haha it actually can't be anything else at the moment, removing ambiguous comment.

File: src/main/java/application/FormController.java (hunk 2 above)
Comment (wpride) [inline src/main/java/application/FormController.java]: Good call, changed this during testing and needs to go back.

File: src/main/java/repo/impl/PostgresMenuSessionRepo.java (hunk 5 above)
Comment (wpride) [inline src/main/java/repo/impl/PostgresMenuSessionRepo.java]: Yes, we return on line 51 otherwise.

**Comment by wpride** [comment]
@benrudolph agreed, these share a lot of code now. I think I'd rather address that in a later PR - I could imagine going a few different ways with it now (interface, abstract class, or merging the two completely)

File: src/main/java/repo/impl/PostgresMenuSessionRepo.java (hunk 5 above)
Comment (benrudolph) [inline src/main/java/repo/impl/PostgresMenuSessionRepo.java]: 👍 missed that

**Comment by benrudolph** [comment]
ok sounds good

**Pull request merged**